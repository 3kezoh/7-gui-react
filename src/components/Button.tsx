import { cn } from "../utils";

type ButtonProps = React.ComponentPropsWithoutRef<"button">;

export function Button(props: ButtonProps) {
  const { className, disabled, ...rest } = props;

  const buttonClassName = cn("border border-black px-2", className, {
    "cursor-not-allowed": disabled,
  });

  return (
    <button
      className={buttonClassName}
      disabled={disabled}
      type="button"
      {...rest}
    />
  );
}
