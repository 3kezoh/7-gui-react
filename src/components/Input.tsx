import { cn } from "../utils";

type InputProps = React.ComponentPropsWithoutRef<"input">;

export function Input(props: InputProps) {
  const { className, name, ...rest } = props;
  const inputClassName = cn("border border-black w-full max-w-xs", className);

  return (
    <input
      className={inputClassName}
      type="text"
      id={name}
      name={name}
      {...rest}
    />
  );
}
