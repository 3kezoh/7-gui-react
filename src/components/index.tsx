export * from "./Button";
export * from "./CRUD";
export * from "./Counter";
export * from "./FlightBooker";
export * from "./Input";
export * from "./TemperatureConverter";
export * from "./Timer";
