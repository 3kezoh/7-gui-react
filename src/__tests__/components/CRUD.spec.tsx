import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CRUD } from "../../components";

describe("CRUD", () => {
  it("should be in the document", () => {
    render(<CRUD />);

    const crud = screen.getByTestId("crud");

    expect(crud).toBeInTheDocument();
  });

  it("should create a user", async () => {
    const user = userEvent.setup();

    render(<CRUD />);

    const nameInput = screen.getByLabelText(/^name:/i);
    const surnameInput = screen.getByLabelText(/surname:/i);
    const createButton = screen.getByRole("button", { name: /create/i });

    await user.type(nameInput, "Leone");
    await user.type(surnameInput, "Abbacchio");
    await user.click(createButton);

    const leone = screen.getByRole("button", { name: /leone, abbacchio/i });

    expect(leone).toBeInTheDocument();
  });

  it("should update a user", async () => {
    const user = userEvent.setup();

    const users = [
      {
        id: "64fe0df3-9291-480b-a3b4-bb060d933592",
        name: "Leone",
        surname: "Abbacchio",
      },
      {
        id: "ad8bce6b-b733-40a6-9114-ea69816627d5",
        name: "Gyro",
        surname: "Zepelli",
      },
    ];

    render(<CRUD initialUsers={users} />);

    const nameInput = screen.getByLabelText(/^name:/i);
    const surnameInput = screen.getByLabelText(/surname:/i);
    const updateButton = screen.getByRole("button", { name: /update/i });
    const leone = screen.getByRole("button", { name: /leone, abbacchio/i });

    await user.click(leone);
    await user.clear(nameInput);
    await user.type(nameInput, "Enrico");
    await user.clear(surnameInput);
    await user.type(surnameInput, "Pucci");
    await user.click(updateButton);

    const enrico = screen.getByRole("button", {
      name: /enrico, pucci/i,
    });

    expect(enrico).toBeInTheDocument();
  });

  it("should delete a user", async () => {
    const user = userEvent.setup();

    const users = [
      {
        id: "64fe0df3-9291-480b-a3b4-bb060d933592",
        name: "Leone",
        surname: "Abbacchio",
      },
    ];

    render(<CRUD initialUsers={users} />);

    const deleteButton = screen.getByRole("button", { name: /delete/i });
    const leone = screen.getByRole("button", { name: /leone, abbacchio/i });

    await user.click(leone);
    await user.click(deleteButton);

    expect(leone).not.toBeInTheDocument();
  });

  it("should disable the update and delete buttons when no person is selected.", async () => {
    const users = [
      {
        id: "64fe0df3-9291-480b-a3b4-bb060d933592",
        name: "Leone",
        surname: "Abbacchio",
      },
    ];

    render(<CRUD initialUsers={users} />);

    const updateButton = screen.getByRole("button", { name: /update/i });
    const deleteButton = screen.getByRole("button", { name: /delete/i });

    expect(updateButton).toBeDisabled();
    expect(deleteButton).toBeDisabled();
  });

  it("should unselect a user", async () => {
    const user = userEvent.setup();

    const users = [
      {
        id: "64fe0df3-9291-480b-a3b4-bb060d933592",
        name: "Leone",
        surname: "Abbacchio",
      },
    ];

    render(<CRUD initialUsers={users} />);

    const leone = screen.getByRole("button", {
      name: /leone, abbacchio/i,
    });

    await user.dblClick(leone);

    expect(leone).toHaveClass("bg-white");
  });

  it("should filter users by their surname", async () => {
    const user = userEvent.setup();

    const users = [
      {
        id: "64fe0df3-9291-480b-a3b4-bb060d933592",
        name: "Leone",
        surname: "Abbacchio",
      },
      {
        id: "ad8bce6b-b733-40a6-9114-ea69816627d5",
        name: "Gyro",
        surname: "Zepelli",
      },
      {
        id: "4b3ac0e4-f21a-46f0-93d9-d9ce2dd96f73",
        name: "Enrico",
        surname: "Pucci",
      },
    ];

    render(<CRUD initialUsers={users} />);

    const prefixInput = screen.getByLabelText(/filter prefix:/i);

    await user.type(prefixInput, "Z");

    const filteredUsers = screen.getAllByRole("button", {
      name: /,/,
    });

    expect(filteredUsers).toHaveLength(1);
  });

  it("should match the snapshot", () => {
    const { asFragment } = render(<CRUD />);

    expect(asFragment()).toMatchSnapshot();
  });
});
