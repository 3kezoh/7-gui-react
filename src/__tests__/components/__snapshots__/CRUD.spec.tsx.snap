// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CRUD > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-2 gap-2 w-fit border border-black p-4"
    data-testid="crud"
  >
    <div
      class="flex gap-2"
    >
      <label
        for="prefix"
      >
        Filter prefix:
      </label>
      <input
        class="border border-black w-full max-w-xs flex-1"
        id="prefix"
        name="prefix"
        type="text"
        value=""
      />
    </div>
    <ul
      class="row-start-2 m-0 p-0 border border-black"
    />
    <div
      class="row-start-2 grid grid-cols-[max-content_1fr] gap-2 h-fit"
    >
      <label
        for="name"
      >
        Name:
      </label>
      <input
        class="border border-black w-full max-w-xs"
        id="name"
        name="name"
        type="text"
        value=""
      />
      <label
        for="surname"
      >
        Surname:
      </label>
      <input
        class="border border-black w-full max-w-xs"
        id="surname"
        name="surname"
        type="text"
        value=""
      />
    </div>
    <div
      class="col-span-2 flex gap-2"
    >
      <button
        class="border border-black px-2"
        type="button"
      >
        Create
      </button>
      <button
        class="border border-black px-2 cursor-not-allowed"
        disabled=""
        type="button"
      >
        Update
      </button>
      <button
        class="border border-black px-2 cursor-not-allowed"
        disabled=""
        type="button"
      >
        Delete
      </button>
    </div>
  </div>
</DocumentFragment>
`;
